Unable to upgrade from 4.x version to the release branch

We are getting a SQL error when you try to upgrade from Dotcms_4.x to the release branch, seems we are trying to use the "cluster_salt" column before run the task that creates it. 

```
11:41:44.122  INFO  caffine.CaffineCache - ***	 Building Cache : dotcmslayoutcache, size:1000,Concurrency:32
11:41:44.185  INFO  startup.StartupTasksExecutor - Not Running Startup Tasks: Task00007RemoveSitesearchQuartzJob
11:41:44.185  INFO  startup.StartupTasksExecutor - Running Startup Tasks : Task00040CheckAnonymousUser
com.dotmarketing.exception.DotRuntimeException:  Error computing cluster data.
	at com.dotcms.security.apps.SecretsKeyStoreHelper.getSecretsStore(SecretsKeyStoreHelper.java:151)
	at com.dotcms.security.apps.SecretsKeyStoreHelper.lambda$3(SecretsKeyStoreHelper.java:232)
	at com.rainerhahnekamp.sneakythrow.Sneaky.lambda$sneaked$8(Sneaky.java:223)
	at com.dotcms.security.apps.SecretsKeyStoreHelper.size(SecretsKeyStoreHelper.java:232)
	at com.dotmarketing.startup.runalways.Task00050LoadAppsSecrets.forceRun(Task00050LoadAppsSecrets.java:83)
	at com.dotmarketing.startup.StartupTasksExecutor.executeStartUpTasks(StartupTasksExecutor.java:130)
	at com.liferay.portal.servlet.MainServlet.init_aroundBody0(MainServlet.java:121)
	at com.liferay.portal.servlet.MainServlet$AjcClosure1.run(MainServlet.java:1)
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149)
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42)
	at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29)
	at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41)
	at com.liferay.portal.servlet.MainServlet.init(MainServlet.java:100)
	at org.apache.catalina.core.StandardWrapper.initServlet(StandardWrapper.java:1144)
	at org.apache.catalina.core.StandardWrapper.loadServlet(StandardWrapper.java:1091)
	at org.apache.catalina.core.StandardWrapper.load(StandardWrapper.java:983)
	at org.apache.catalina.core.StandardContext.loadOnStartup(StandardContext.java:4978)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5290)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:150)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:754)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:730)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:734)
	at org.apache.catalina.startup.HostConfig.deployDirectory(HostConfig.java:1140)
	at org.apache.catalina.startup.HostConfig$DeployDirectory.run(HostConfig.java:1875)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.dotmarketing.exception.DotRuntimeException:  Error computing cluster data.
	at com.dotcms.security.apps.SecretsKeyStoreHelper.createStoreIfNeeded(SecretsKeyStoreHelper.java:116)
	at com.dotcms.security.apps.SecretsKeyStoreHelper.getSecretsStore(SecretsKeyStoreHelper.java:133)
	... 28 more
Caused by: com.dotmarketing.exception.DotRuntimeException:  Error computing cluster data.
	at com.dotcms.enterprise.cluster.ClusterFactory.a(SourceFile:220)
	at io.vavr.Lazy.computeValue(Lazy.java:161)
	at io.vavr.Lazy.get(Lazy.java:155)
	at com.dotcms.enterprise.cluster.ClusterFactory.getClusterData(SourceFile:147)
	at com.dotcms.enterprise.cluster.ClusterFactory.getClusterSalt(SourceFile:134)
	at com.dotcms.security.apps.SecretsKeyStoreHelper.lambda$1(SecretsKeyStoreHelper.java:94)
	at com.dotcms.security.apps.SecretsKeyStoreHelper.createStoreIfNeeded(SecretsKeyStoreHelper.java:108)
	... 29 more
Caused by: com.dotmarketing.exception.DotDataException: ERROR: column "cluster_salt" does not exist
  Position: 20{
  "SQL": ["select cluster_id, cluster_salt from dot_cluster"],
  "maxRows": [-1],
  "offest": [0]
}
	at com.dotmarketing.common.db.DotConnect.loadResult(DotConnect.java:273)
	at com.dotmarketing.common.db.DotConnect.loadObjectResults(DotConnect.java:925)
	at com.dotcms.enterprise.cluster.ClusterFactory.a(SourceFile:168)
	... 35 more
Caused by: org.postgresql.util.PSQLException: ERROR: column "cluster_salt" does not exist
  Position: 20
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2497)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2233)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:310)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:446)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:370)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:311)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:297)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:274)
	at org.postgresql.jdbc.PgStatement.executeQuery(PgStatement.java:225)
	at com.zaxxer.hikari.pool.ProxyStatement.executeQuery(ProxyStatement.java:111)
	at com.zaxxer.hikari.pool.HikariProxyStatement.executeQuery(HikariProxyStatement.java)
	at com.dotmarketing.common.db.DotConnect.executeQuery(DotConnect.java:646)
	at com.dotmarketing.common.db.DotConnect.executeQuery(DotConnect.java:532)
	at com.dotmarketing.common.db.DotConnect.loadResult(DotConnect.java:271)
	... 37 more
```

**Expected behavior**
We need be able to upgrade without problems

--- a/dotCMS/src/main/java/com/dotmarketing/startup/runalways/Task00006CreateSystemLayout.java
+++ b/dotCMS/src/main/java/com/dotmarketing/startup/runalways/Task00006CreateSystemLayout.java
@@ -1,5 +1,6 @@
 package com.dotmarketing.startup.runalways;
 
+import com.dotmarketing.startup.runonce.Task04355SystemEventAddServerIdColumn;
 import java.util.ArrayList;
 import java.util.List;
 
@@ -44,6 +45,15 @@ public class Task00006CreateSystemLayout implements StartupTask {
 	}
 
 	public boolean forceRun() {
+		try {
+			final Task04355SystemEventAddServerIdColumn systemEventAddServerIdColumn = new Task04355SystemEventAddServerIdColumn();
+			if (systemEventAddServerIdColumn.forceRun()) {
+				systemEventAddServerIdColumn.executeUpgrade();
+			}
+		} catch (Exception e) {
+			Logger.error(Task00006CreateSystemLayout.class,
+					"Error Applying upgrade task from Task00006CreateSystemLayout.", e);
+		}
 		return true;
 	}
 

--- a/dotCMS/src/main/java/com/dotmarketing/startup/runalways/Task00050LoadAppsSecrets.java
+++ b/dotCMS/src/main/java/com/dotmarketing/startup/runalways/Task00050LoadAppsSecrets.java
@@ -11,6 +11,7 @@ import static com.dotcms.security.apps.AppsUtil.validateForSave;
 import static com.dotmarketing.util.UtilMethods.isSet;
 import static com.liferay.util.StringPool.BLANK;
 
+import com.dotcms.enterprise.cluster.ClusterFactory.ClusterData;
 import com.dotcms.security.apps.AppDescriptor;
 import com.dotcms.security.apps.AppDescriptorHelper;
 import com.dotcms.security.apps.AppSecrets;
@@ -22,6 +23,8 @@ import com.dotmarketing.common.db.DotConnect;
 import com.dotmarketing.exception.DotDataException;
 import com.dotmarketing.exception.DotRuntimeException;
 import com.dotmarketing.startup.StartupTask;
+import com.dotmarketing.startup.runonce.Task04355SystemEventAddServerIdColumn;
+import com.dotmarketing.startup.runonce.Task05350AddDotSaltClusterColumn;
 import com.dotmarketing.util.Config;
 import com.dotmarketing.util.Logger;
 import com.google.common.annotations.VisibleForTesting;
@@ -80,6 +83,7 @@ public class Task00050LoadAppsSecrets implements StartupTask {
 
     @Override
     public boolean forceRun() {
+        addDotSaltClusterColumnIfNeeded();
         return keyStoreHelper.size() == 0 && isSet(Config.getStringProperty(APPS_IMPORT_EXPORT_DEFAULT_PASSWORD, BLANK)) ;
     }
 
@@ -151,6 +155,29 @@ public class Task00050LoadAppsSecrets implements StartupTask {
         } catch (Exception e) {
             Logger.error(Task00050LoadAppsSecrets.class, "Error Importing AppSecret from starter.",
                     e);
+        }
+    }
+
+    /**
+     * This Task attempts to create a security repo that depends on a call to {@link ClusterData#getClusterSalt()}
+     * This creates a conflict When upgrading from 4.x since back then we didn't have the column `cluster_salt`
+     * For which we need to call the upgrade task from here
+     */
+    private void addDotSaltClusterColumnIfNeeded() {
+        try {
+            final Task05350AddDotSaltClusterColumn task05350AddDotSaltClusterColumn = new Task05350AddDotSaltClusterColumn();
+            if (task05350AddDotSaltClusterColumn.forceRun()) {
+                task05350AddDotSaltClusterColumn.executeUpgrade();
+            }
+
+            final Task04355SystemEventAddServerIdColumn systemEventAddServerIdColumn = new Task04355SystemEventAddServerIdColumn();
+            if (systemEventAddServerIdColumn.forceRun()) {
+                systemEventAddServerIdColumn.executeUpgrade();
+            }
+
+        } catch (Exception e) {
+            Logger.error(Task00050LoadAppsSecrets.class,
+                    "Error Applying upgrade task from Task00050LoadAppsSecrets.", e);
         }
     }
 

--- a/dotCMS/src/main/java/com/dotmarketing/startup/runalways/Task00050LoadAppsSecrets.java
+++ b/dotCMS/src/main/java/com/dotmarketing/startup/runalways/Task00050LoadAppsSecrets.java
@@ -11,7 +11,6 @@ import static com.dotcms.security.apps.AppsUtil.validateForSave;
 import static com.dotmarketing.util.UtilMethods.isSet;
 import static com.liferay.util.StringPool.BLANK;
 
-import com.dotcms.enterprise.cluster.ClusterFactory.ClusterData;
 import com.dotcms.security.apps.AppDescriptor;
 import com.dotcms.security.apps.AppDescriptorHelper;
 import com.dotcms.security.apps.AppSecrets;

--- a/dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task04355SystemEventAddServerIdColumn.java
+++ b/dotCMS/src/main/java/com/dotmarketing/startup/runonce/Task04355SystemEventAddServerIdColumn.java
@@ -3,6 +3,7 @@
 package com.dotmarketing.startup.runonce;
 
 import com.dotmarketing.common.db.DotConnect;
+import com.dotmarketing.common.db.DotDatabaseMetaData;
 import com.dotmarketing.db.DbConnectionFactory;
 import com.dotmarketing.db.HibernateUtil;
 import com.dotmarketing.exception.DotDataException;
@@ -43,7 +44,12 @@ public class Task04355SystemEventAddServerIdColumn extends AbstractJDBCStartupTa
 
     @Override
     public boolean forceRun() {
-        return Boolean.TRUE;
+        try {
+            return !new DotDatabaseMetaData().hasColumn("system_event", "server_id");
+        } catch (SQLException e) {
+            Logger.error(this, e.getMessage(),e);
+            return false;
+        }
     }
 
     @Override

